Copy server error processing from driver

mdehoog opened the pull request:
We weren't calling `server.ProcessError` on any round trip errors. This means that any connection errors wouldn't mark the server as unknown, extending failures.

This PR adds the `server.ProcessError` call, and duplicates the wrapping of the error in a `driver.Error` just as `driver.Operation` does in the Go driver.

divjotarora commented:
Hi @mdehoog,

I work on the Go Driver at MongoDB and saw this PR open up. I agree that this could cause prolonged failures and requests to nodes that aren't available. A few small suggestions:

1. If it's possible to encounter a network timeout in the `mongobetween` proxy, I'd recommend bumping the driver version to v1.3.4 to pull in the change for https://jira.mongodb.org/browse/GODRIVER-1620. Before that fix, the driver was marking the server as Unknown and clearing its connection pool for all network errors, including timeouts. This caused a lot of connection churn if a timeout happened and the pool had a lot of connections. I don't think this affects the `mongobetween` proxy because the context used for the `WriteWireMessage`/`ReadWireMessage` calls is a cancel-able version of `context.Background()` and I don't see any calls to `ClientOptions.SetSocketTimeout` in the code, so network timeouts shouldn't be possible, but I figured I'd bring it up just in case.

2. `topology.Server.ProcessError` accounts for some special command errors like `NotMaster`, `NodeIsRecovering`, `ShutdownInProgress`, etc, in addition to connection errors. Is this handling important for the proxy?

3. Is it possible to write a test for this? The driver tests this using connection pool monitoring in https://github.com/mongodb/mongo-go-driver/blob/v1.3.4/mongo/integration/sdam_error_handling_test.go.

-- Divjot

divjotarora commented on mongo/mongo.go:
```diff
@@ -181,6 +181,10 @@ func (m *Mongo) RoundTrip(msg *Message) (*Message, error) {
 	wm, err := m.roundTrip(conn, msg.Wm)
 	if err != nil {
+		// see https://github.com/mongodb/mongo-go-driver/blob/v1.3.2/x/mongo/driver/operation.go#L368-L370
+		if ep, ok := server.(driver.ErrorProcessor); ok {
```
I'd recommend a stricter check. The driver has this conditional check because the deployment types used by the SDAM monitoring routines and during connection handshakes don't implement `ErrorProcessor`. However, the proxy should always be using a server of type `*topology.Server`, which must implement `ErrorProcessor`, so I'd recommend either failing or at least logging an error if `ok` is false.

mdehoog replied:
Good suggestion, added a warning log.

mdehoog commented:
Hey @divjotarora, thanks for the review and comment, glad to have you looking at `mongobetween`!

1. I noticed that the `topology.ConnectionError` wasn't being unwrapped correctly in `unwrapConnectionError`, but hadn't realized that this was fixed in `v1.3.4`. Thanks for the heads up. Updated the driver to 1.3.4 in #9.
2. Yes I think it would be helpful to have this additional handling 👍 
3. Added a test, thanks for the reference.

divjotarora commented:
Sorry if this wasn't clear, but point (2) in my previous comment was trying to say that the `*Server.ProcessError` will handle things besides connection errors, but it's use in this PR will not. That is because it's called right after `roundTrip` but before `Decode`, so it will only ever be called for connection errors. Contrast this with the driver's `Operation.roundTrip` function (https://github.com/mongodb/mongo-go-driver/blob/v1.3.4/x/mongo/driver/operation.go#L532), which does the network roundtrip, but also decodes the result and extracts the error.

mdehoog commented:
Makes sense. I'll look at duplicating any relevant `extractError` functionality as a follow-up.
